File: course_2/module_3/examples/binary_search_tree.go
package main

import (
	"errors"
	"fmt"
)

// TreeNode represents a node in the binary search tree.
type TreeNode struct {
	Key    int
	Left   *TreeNode
	Right  *TreeNode
	Parent *TreeNode
	Size   int // Size of the subtree, including this node.
}

// BinarySearchTree represents the binary search tree structure.
type BinarySearchTree struct {
	Root *TreeNode
}

// NewTreeNode creates a new tree node.
func NewTreeNode(key int) *TreeNode {
	return &TreeNode{
		Key:  key,
		Size: 1, // A single node has size 1.
	}
}

// Insert adds a new key to the BST. Ignores duplicates.
func (bst *BinarySearchTree) Insert(key int) {
	if bst.Search(key) != nil {
		return // Ignore duplicates.
	}
	newNode := NewTreeNode(key)
	if bst.Root == nil {
		bst.Root = newNode
		return
	}
	current := bst.Root
	for current != nil {
		current.Size++ // Update size of the subtree.
		if key < current.Key {
			if current.Left == nil {
				current.Left = newNode
				newNode.Parent = current
				bst.updateSize(current)
				return
			}
			current = current.Left
		} else {
			if current.Right == nil {
				current.Right = newNode
				newNode.Parent = current
				bst.updateSize(current)
				return
			}
			current = current.Right
		}
	}
}

// Search finds a node with the given key in the BST.
func (bst *BinarySearchTree) Search(key int) *TreeNode {
	current := bst.Root
	for current != nil {
		if key == current.Key {
			return current
		} else if key < current.Key {
			current = current.Left
		} else {
			current = current.Right
		}
	}
	return nil
}

// Minimum finds the node with the smallest key in the BST or subtree.
func (bst *BinarySearchTree) Minimum(node *TreeNode) *TreeNode {
	if node == nil {
		node = bst.Root
	}
	for node != nil && node.Left != nil {
		node = node.Left
	}
	return node
}

// Maximum finds the node with the largest key in the BST or subtree.
func (bst *BinarySearchTree) Maximum(node *TreeNode) *TreeNode {
	if node == nil {
		node = bst.Root
	}
	for node != nil && node.Right != nil {
		node = node.Right
	}
	return node
}

// Successor finds the next largest key after the given key.
func (bst *BinarySearchTree) Successor(key int) (*int, error) {
	node := bst.Search(key)
	if node == nil {
		return nil, errors.New("key not found")
	}
	if node.Right != nil {
		min := bst.Minimum(node.Right)
		return &min.Key, nil
	}
	current := node.Parent
	for current != nil && node == current.Right {
		node = current
		current = current.Parent
	}
	if current == nil {
		return nil, errors.New("no successor found")
	}
	return &current.Key, nil
}

// Predecessor finds the next smaller key before the given key.
func (bst *BinarySearchTree) Predecessor(key int) (*int, error) {
	node := bst.Search(key)
	if node == nil {
		return nil, errors.New("key not found")
	}
	if node.Left != nil {
		max := bst.Maximum(node.Left)
		return &max.Key, nil
	}
	current := node.Parent
	for current != nil && node == current.Left {
		node = current
		current = current.Parent
	}
	if current == nil {
		return nil, errors.New("no predecessor found")
	}
	return &current.Key, nil
}

// Delete removes a node with the given key from the BST.
func (bst *BinarySearchTree) Delete(key int) error {
	node := bst.Search(key)
	if node == nil {
		return errors.New("key not found")
	}
	if node.Left == nil && node.Right == nil { // Case 1: No children.
		bst.transplant(node, nil)
	} else if node.Right == nil { // Case 2: One child (left).
		bst.transplant(node, node.Left)
	} else if node.Left == nil { // Case 2: One child (right).
		bst.transplant(node, node.Right)
	} else { // Case 3: Two children.
		successor := bst.Minimum(node.Right)
		if successor.Parent != node {
			bst.transplant(successor, successor.Right)
			successor.Right = node.Right
			successor.Right.Parent = successor
		}
		bst.transplant(node, successor)
		successor.Left = node.Left
		successor.Left.Parent = successor
		successor.Size = node.Size
	}
	bst.updateSize(node.Parent) // Update sizes up the tree.
	return nil
}

// transplant replaces one subtree with another.
func (bst *BinarySearchTree) transplant(u, v *TreeNode) {
	if u.Parent == nil {
		bst.Root = v
	} else if u == u.Parent.Left {
		u.Parent.Left = v
	} else {
		u.Parent.Right = v
	}
	if v != nil {
		v.Parent = u.Parent
	}
}

// updateSize updates the sizes of all ancestor nodes.
func (bst *BinarySearchTree) updateSize(node *TreeNode) {
	for node != nil {
		node.Size = 1
		if node.Left != nil {
			node.Size += node.Left.Size
		}
		if node.Right != nil {
			node.Size += node.Right.Size
		}
		node = node.Parent
	}
}

// InOrderTraversal returns the keys in sorted order.
func (bst *BinarySearchTree) InOrderTraversal() []int {
	var result []int
	var inOrder func(node *TreeNode)
	inOrder = func(node *TreeNode) {
		if node == nil {
			return
		}
		inOrder(node.Left)
		result = append(result, node.Key)
		inOrder(node.Right)
	}
	inOrder(bst.Root)
	return result
}

// Main function to demonstrate the BinarySearchTree operations.
func main() {
	bst := &BinarySearchTree{}

	// Insert keys.
	bst.Insert(15)
	bst.Insert(10)
	bst.Insert(20)
	bst.Insert(8)
	bst.Insert(12)

	// In-order traversal.
	fmt.Println("In-order Traversal:", bst.InOrderTraversal())

	// Find minimum and maximum.
	fmt.Println("Minimum:", bst.Minimum(nil).Key)
	fmt.Println("Maximum:", bst.Maximum(nil).Key)

	// Find successor and predecessor.
	successor, _ := bst.Successor(10)
	fmt.Println("Successor of 10:", *successor)
	predecessor, _ := bst.Predecessor(15)
	fmt.Println("Predecessor of 15:", *predecessor)

	// Delete a node and display updated tree.
	_ = bst.Delete(10)
	fmt.Println("In-order Traversal after deleting 10:", bst.InOrderTraversal())
}

File: course_2/module_3/examples/binary_search_tree.py
class TreeNode:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
        self.parent = None
        self.size = 1  # Size of the subtree, including this node.


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, key):
        """Insert a key into the tree, ignoring duplicates."""
        if self.search(key):  # Skip if the key already exists.
            return
        new_node = TreeNode(key)
        if not self.root:
            self.root = new_node  # Set root if tree is empty.
            return
        current = self.root
        while current:
            current.size += 1  # Update the size of each ancestor node.
            if key < current.key:
                if not current.left:
                    current.left = new_node
                    new_node.parent = current
                    return
                current = current.left
            else:
                if not current.right:
                    current.right = new_node
                    new_node.parent = current
                    return
                current = current.right

    def search(self, key):
        """Search for a key in the tree and return its node, or None if not found."""
        current = self.root
        while current and current.key != key:
            current = current.left if key < current.key else current.right
        return current

    def minimum(self, node=None):
        """Return the node with the minimum key in the tree or subtree."""
        if node is None:
            node = self.root
        while node and node.left:
            node = node.left  # Traverse left until null.
        return node

    def maximum(self, node=None):
        """Return the node with the maximum key in the tree or subtree."""
        if node is None:
            node = self.root
        while node and node.right:
            node = node.right  # Traverse right until null.
        return node

    def successor(self, key):
        """Return the key of the next largest element after the given key."""
        node = self.search(key)
        if not node:
            return None
        if node.right:
            return self.minimum(node.right).key  # The successor is the minimum in the right subtree.
        current = node.parent
        while current and node == current.right:
            node = current
            current = current.parent
        return current.key if current else None  # Return successor or None if not found.

    def predecessor(self, key):
        """Return the key of the previous smaller element before the given key."""
        node = self.search(key)
        if not node:
            return None
        if node.left:
            return self.maximum(node.left).key  # The predecessor is the maximum in the left subtree.
        current = node.parent
        while current and node == current.left:
            node = current
            current = current.parent
        return current.key if current else None  # Return predecessor or None if not found.

    def delete(self, key):
        """Delete the node with the given key from the tree."""
        node = self.search(key)
        if not node:
            raise ValueError(f"Key {key} not found in the tree.")

        # Update the sizes of ancestor nodes.
        current = node
        while current:
            current.size -= 1
            current = current.parent

        if not node.left and not node.right:  # Case 1: No children.
            self._transplant(node, None)
        elif not node.right:  # Case 2: One child (left).
            self._transplant(node, node.left)
        elif not node.left:  # Case 2: One child (right).
            self._transplant(node, node.right)
        else:  # Case 3: Two children.
            successor = self.minimum(node.right)
            if successor.parent != node:
                self._transplant(successor, successor.right)
                successor.right = node.right
                successor.right.parent = successor
            self._transplant(node, successor)
            successor.left = node.left
            successor.left.parent = successor
            successor.size = node.size  # Update the size of the new root.

    def _transplant(self, u, v):
        """Replace the subtree rooted at u with the subtree rooted at v."""
        if not u.parent:
            self.root = v  # If u is root, replace root.
        elif u == u.parent.left:
            u.parent.left = v
        else:
            u.parent.right = v
        if v:
            v.parent = u.parent  # Update parent pointer of v.

    def select(self, i):
        """Return the key of the i-th smallest element in the tree (1-based index)."""
        def _select(node, i):
            if not node:
                return None
            left_size = node.left.size if node.left else 0
            if i == left_size + 1:
                return node
            elif i <= left_size:
                return _select(node.left, i)
            else:
                return _select(node.right, i - left_size - 1)
        node = _select(self.root, i)
        return node.key if node else None

    def rank(self, key):
        """Return the number of keys less than or equal to the given key."""
        node = self.search(key)
        if not node:
            return None
        rank = 1
        if node.left:
            rank += node.left.size
        while node != self.root:
            if node == node.parent.right:
                rank += 1
                if node.parent.left:
                    rank += node.parent.left.size
            node = node.parent
        return rank

    def inorder_traversal(self):
        """Return a list of keys in sorted order."""
        result = []

        def _inorder(node):
            if not node:
                return
            _inorder(node.left)
            result.append(node.key)
            _inorder(node.right)

        _inorder(self.root)
        return result


if __name__ == "__main__":
    # Example usage of BinarySearchTree.
    bst = BinarySearchTree()
    bst.insert(15)
    bst.insert(10)
    bst.insert(20)
    bst.insert(8)
    bst.insert(12)

    print("In-order Traversal:", bst.inorder_traversal())
    print("Minimum:", bst.minimum().key)
    print("Maximum:", bst.maximum().key)
    print("Rank of 12:", bst.rank(12))
    print("Select 3rd smallest:", bst.select(3))
    print("Successor of 10:", bst.successor(10))
    print("Predecessor of 15:", bst.predecessor(15))

    bst.delete(10)
    print("In-order Traversal after deleting 10:", bst.inorder_traversal())
